presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DetermineJoinDistributionType.java

```diff
@@ -99,8 +99,17 @@ private PlanNode getCostBasedJoin(JoinNode joinNode, Context context)
     {
         List<PlanNodeWithCost> possibleJoinNodes = new ArrayList<>();
 
-        addJoinsWithDifferentDistributions(joinNode, possibleJoinNodes, context);
-        addJoinsWithDifferentDistributions(joinNode.flipChildren(), possibleJoinNodes, context);
+        // don't consider right inequality joins because they execute on a single node.
+        if (joinNode.getCriteria().isEmpty() && joinNode.getType() == LEFT) {
+            addJoinsWithDifferentDistributions(joinNode, possibleJoinNodes, context);
+        }
+        else if (joinNode.getCriteria().isEmpty() && joinNode.getType() == RIGHT) {
+            addJoinsWithDifferentDistributions(joinNode.flipChildren(), possibleJoinNodes, context);
+        }
+        else {
+            addJoinsWithDifferentDistributions(joinNode, possibleJoinNodes, context);
+            addJoinsWithDifferentDistributions(joinNode.flipChildren(), possibleJoinNodes, context);
+        }
 
         if (possibleJoinNodes.stream().anyMatch(result -> result.getCost().hasUnknownComponents()) || possibleJoinNodes.isEmpty()) {
```

Comment:
Now you need to check for `possibleJoinNodes.isEmpty() &&`

Reply:
we already do.  Do you mean to put it first?

Reply:
Sorry, i missed that. You can put this first.